pkg/services/sqlstore/sqlstore.go

```diff
@@ -425,6 +442,118 @@ func (ss *SQLStore) readConfig() error {
+	subsystem := "database"
+
+	s.metrics.maxOpenConnections = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: namespace,
+		Subsystem: subsystem,
+		Name:      "conn_max_open",
+		Help:      "Maximum number of open connections to the database",
+	})
+
+	s.metrics.openConnections = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: namespace,
+		Subsystem: subsystem,
+		Name:      "conn_open",
+		Help:      "The number of established connections both in use and idle",
+	})
+
+	s.metrics.inUse = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: namespace,
+		Subsystem: subsystem,
+		Name:      "conn_in_use",
+		Help:      "The number of connections currently in use",
+	})
+
+	s.metrics.idle = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: namespace,
+		Subsystem: subsystem,
+		Name:      "conn_idle",
+		Help:      "The number of idle connections",
+	})
+
+	s.metrics.waitCount = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: namespace,
+		Subsystem: subsystem,
+		Name:      "conn_wait_count_total",
+		Help:      "The total number of connections waited for",
+	})
+
+	s.metrics.waitDuration = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: namespace,
+		Subsystem: subsystem,
+		Name:      "conn_wait_duration_seconds",
+		Help:      "The total time blocked waiting for a new connection",
+	})
+
+	s.metrics.maxIdleClosed = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: namespace,
+		Subsystem: subsystem,
+		Name:      "conn_max_idle_closed_total",
+		Help:      "The total number of connections closed due to SetMaxIdleConns",
+	})
+
+	s.metrics.maxIdleTimeClosed = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: namespace,
+		Subsystem: subsystem,
+		Name:      "conn_max_idle_closed_seconds",
+		Help:      "The total number of connections closed due to SetConnMaxIdleTime",
+	})
+
+	s.metrics.maxLifetimeClosed = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: namespace,
+		Subsystem: subsystem,
+		Name:      "conn_max_lifetime_closed_total",
+		Help:      "The total number of connections closed due to SetConnMaxLifetime",
+	})
+}
+
+// collectDBStats instruments connections stats from the database.
+func (s *SQLStore) collectDBstats() {
+	dbstats := s.engine.DB().Stats()
+	s.metrics.maxOpenConnections.Set(float64(dbstats.MaxOpenConnections))
+	s.metrics.openConnections.Set(float64(dbstats.MaxOpenConnections))
+	s.metrics.inUse.Set(float64(dbstats.InUse))
+	s.metrics.idle.Set(float64(dbstats.Idle))
+
+	s.metrics.waitCount.Add(float64(dbstats.WaitCount))
+	s.metrics.waitDuration.Add(float64(dbstats.WaitDuration / time.Second))
+	s.metrics.maxIdleClosed.Add(float64(dbstats.MaxIdleClosed))
+	s.metrics.maxIdleTimeClosed.Add(float64(dbstats.MaxIdleTimeClosed))
+	s.metrics.maxLifetimeClosed.Add(float64(dbstats.MaxLifetimeClosed))
+}
+
+// Collect implements Prometheus.Collector.
+func (s *SQLStore) Collect(ch chan<- prometheus.Metric) {
+	s.collectDBstats()
```

Comment:
I decided to collect the stats here instead of creating another go routine just for collecting stats.